### DISCUSSION TITLE
Image affine  transformations and overlaying it over a background image

### DISCUSSION THREAD

**Discussion by 2246sam20** [ImageMagick/ImageMagick#7543] [Help]
### ImageMagick version

 7.1.1-37

### Operating system, version and so on

Windows 11

### Description

Hello! 
I am trying to apply translate, scale, rotate (affine transformations ) to an image. After applying the transformations i want to overlay this image over a background image.
Magick::Image overlayImage -> overlay Image
Magick::Image backgroundImage-> background Image

I am using Magick::DrawableList for pushing the transformation drawable objects into it
like Magick::DrawableTranslation  , Magick::DrawableRotation and so on for other transformations 

overlayImage.draw(drawList); -> applying transformations 

backgroundImage.composite(overlayImage, 0, 0, Magick::OverCompositeOp);  - >overlaying image
There's no change to the overlay Image after the transformations, it is still  appearing at the (0,0) on the background Image even after some non zero tx and ty. I know that in composite method we are passing x,y  coords for overlay Image but the transformations should shift the overlay Image.

But the same transformation method I am trying above  are working for Drawable Objects like  Magick::DrawableEllipse 
image.draw(drawList);   // drawList contains objects like Magick::DrawableTranslation , Magick::DrawableRotation
where image object is a Magick::Image camvas where we are rendering DrawableEllipse  and other objects like this.

Any help will be much appreciated. I tried other methods like Magick::Drawable Affine but no luck.
P.S. - The Transformation order matters, so how to handle it also  while applying them to the overlay Image.

P.S.S = I tried with Distort command too using cmd but not results, maybe i am doing something wrong.

**Comment by snibgo** [comment]
You don't show your source code, so we can only guess.

You seem to have set up transformations for drawing an object, but then don't actually draw anything.

If you want to distort an image by scale, rotate and translate, then the obvious method would seem to be `distort`. Then composite that over the background.

**Reply by 2246sam20** [reply]
Thanks for the reply
I did mention that I am drawing like this -> image.draw(drawList); 
Let me know if you need any thing else?

Also for the following transformation translate(-2830.38,127.459) rotate(0) scale(0.996,0.996)
How to use distort command?
Thanks in advance..

**Comment by snibgo** [comment]
I can't debug your program if you don't show it.

For `-distort SRT`, see https://imagemagick.org/script/command-line-options.php#distort

**Reply by 2246sam20** [reply]
```
struct TransformData {
	std::vector<std::string> transformOrder;
	double tx = 0;
	double ty = 0;
	double angle = 0;
	double sx = 1;
	double sy = 1;
	double skewX = 0;
	double skewY = 0;
};

void  applyTransformations(const TransformData& transformData, Magick::DrawableList& drawList) {
	for (const std::string& transformType : transformData.transformOrder) {
		if (transformType == "translate") {
			Magick::DrawableTranslation translation(transformData.tx, transformData.ty);
			drawList.push_back(translation);
		}
		else if (transformType == "rotate") {
			Magick::DrawableRotation rotation(transformData.angle);
			drawList.push_back(rotation);
		}
		else if (transformType == "scale") {
			Magick::DrawableScaling scaling(transformData.sx, transformData.sy);
			drawList.push_back(scaling);
		}
		else if (transformType == "skewX") {
			Magick::DrawableSkewX skewX(transformData.skewX);
			drawList.push_back(skewX);
		}
		else if (transformType == "skewY") {
			Magick::DrawableSkewY skewY(transformData.skewY);
			drawList.push_back(skewY);
		}
	}
}



void renderImage(const Image & image, Magick::Image & backgroundImage) {
	try {
		Magick::Image overlayImage;
		overlayImage.read(image.filePath); // some image 
		if (overlayImage.isValid()) {
			double x = std::stod(image.x);
			double y = std::stod(image.y);
			double width = std::stod(image.width);
			double height = std::stod(image.height);

			overlayImage.resize(Magick::Geometry(width, height));
			TransformData transformData = image.getTransformData(); // some util function that  willl give the transform data 
			
		        Magick::DrawableList drawList;
		        applyTransformations(transformData, drawList);
		        overlayImage.draw(drawList);
	
			backgroundImage.composite(overlayImage, 0, 0, Magick::OverCompositeOp);
		}
		else {
			logError("failed to Load overlay image: " + image.id);
		}
	}
	catch (const std::exception& e) {
		logError( "Error rendering image: " + std::string(e.what()) );
	}
}

```
According to the https://imagemagick.org/script/command-line-options.php#draw   documentation here,
transformation primitives  can be drawn over an image using the -draw option.

**Comment by snibgo** [comment]
My guess was correct: you set up transformations for drawing an object, but then don't actually draw anything.

For example, consider this command:
```
magick toes.png -draw "rotate 30 translate 100,100 rectangle 10,10,30,30" out.png
```
This reads an image, and draws on that image. The drawlist contains three items: `rotate`, `translate` and `rectangle`. The command draws a rectangle that is rotated and translated. If the command was this:
```
magick toes.png -draw "rotate 30 translate 100,100" x.png
```
... then the `-draw` would do nothing.

**Reply by 2246sam20** [reply]
Thanks for the quick reply.
Ahh, you are write I was doing it in a wrong way. But still it doesn't resolve my query 
I want to apply the transformation primitives on the overlay image and then I want to composite it over background image.
I hope my question is clear now.

**Comment by snibgo** [comment]
In a command, we can do this:
```
magick toes.png -draw "rotate 15 image SrcOver 100,100,70,50 r.png" x2.png
```
This reads `r.png`, resizes that to 70x50, translates that to 100,100, rotates that by 15 degrees, and composites that over `toes.png`.

In `Drawable.h`, see ...
```
  DrawableCompositeImage ( double x_, double y_,
                           double width_, double height_,
                           const std::string &filename_ );
```
... and variations of `DrawableCompositeImage`.

**Reply by 2246sam20** [reply]
Thanks @snibgo, I will check it
